### ISSUE TITLE
DTLS performance degradation in v3.2.0 (and later)

### ISSUE THREAD

**Issue by bjosv** [openssl/openssl#26055]
In an application that sends packets over a DTLS channel we noticed a performance degradation when lifting OpenSSL version from `3.1.0` to `3.2.0`. After some measurements on the application we found that it could send `~210k pps` (packets per second) with `OpenSSL 3.1.0`, while we only got `~180k pps` when using `OpenSSL 3.2.0`, which is a 14% degradation.

To get a better understanding of the reason I created a simple reproduction, a simple DTLS [client](https://github.com/bjosv/openssl-experiments/blob/main/tools/dtls-client.c) that just sends 1k of data using `SSL_write_ex`, in a loop, to a simple [server](https://github.com/bjosv/openssl-experiments/blob/main/tools/dtls-server.c).
This simple setup also shows that there is a decrease in performance when using `3.2.3` compared to `3.1.7`,
259k pps in `3.2.3` vs. 282k pps in `3.1.7` (see [details](https://github.com/bjosv/openssl-experiments/blob/main/issues/dtls-performance.md#send-loop-measurements)).

I also built OpenSSL using `./Configure enable-crypto-mdebug` to enable the API `CRYPTO_get_alloc_counts()` which indicates that each DTLS write in `3.2.3` performs 2 mallocs. I don't see any mallocs during a write in `3.1.7`.
Are 2 mallocs expected?

To dig even deeper we performed a bisect to see when the performance took a hit and found following interesting commits:

| Commit | Data | Headline | packet per second |
| --- | --- | --- | --- |
| 602ee1f672 | 2022-10-07 | Use common tls_write_records() even for DTLS | ~210k pps (ok!) |
| 248a9bf21a | 2022-10-13 | Start using WPACKET in the dtls write records code |~200k pps (5% degradation) |
| b9e37f8f57 | 2022-10-13 | Convert dtls_write_records to use standard record layer functions |~180k pps (14% degradation)|

Is this degradation known? I couldn't find any existing issues or notes about it.
Should we improve this somehow, maybe there is plan which I can help out with?

**Comment by t8m** [comment]
> Is this degradation known? I couldn't find any existing issues or notes about it.
> Should we improve this somehow, maybe there is plan which I can help out with?

No, this is not known. Thank you for the report and analysis. We should definitely aim for improving this. Maybe we could use some preallocated buffers & WPACKETs.

**Comment by fwh-dc** [comment]
Does openssl generate performance numbers in the CI that would allow to track performance degradations? If so would it be possible to include the tests you have produced @bjosv ?

**Comment by mattcaswell** [comment]
> Does openssl generate performance numbers in the CI that would allow to track performance degradations?

We have the performance dashboard here:

https://openssl-library.org/performance.html

This is based on tests added to this repository:

https://github.com/openssl/perftools

**Comment by bjosv** [comment]
Interesting, I will check it out to see if I can add some useful testing.
